File: management/api/routers/containers.py
"""
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
/management/api/routers/containers.py

Part of the "n8n_nginx/n8n_management" suite
Version 3.0.0 - January 1st, 2026

Richard J. Sears
richard@n8nmanagement.net
https://github.com/rjsears
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
"""

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Dict, Any

from api.database import get_db
from api.dependencies import get_current_user
from api.services.container_service import ContainerService
from api.schemas.common import SuccessResponse

router = APIRouter()


@router.get("/", response_model=List[Dict[str, Any]])
async def list_containers(
    all: bool = True,
    _=Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List all project containers."""
    service = ContainerService(db)
    return await service.list_containers(all=all)


@router.get("/stats", response_model=List[Dict[str, Any]])
async def get_container_stats(
    _=Depends(get_current_user),
):
    """Get resource usage stats for all containers."""
    service = ContainerService()
    return await service.get_stats()


@router.get("/health")
async def check_containers_health(
    _=Depends(get_current_user),
):
    """Check health of all project containers."""
    service = ContainerService()
    return await service.check_health()


@router.get("/{name}")
async def get_container(
    name: str,
    _=Depends(get_current_user),
):
    """Get container details."""
    service = ContainerService()
    container = await service.get_container(name)

    if not container:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Container not found or not a project container",
        )

    return container


@router.post("/{name}/start", response_model=SuccessResponse)
async def start_container(
    name: str,
    _=Depends(get_current_user),
):
    """Start a container."""
    service = ContainerService()

    try:
        await service.start_container(name)
        return SuccessResponse(message=f"Container {name} started")
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/{name}/stop", response_model=SuccessResponse)
async def stop_container(
    name: str,
    timeout: int = 30,
    _=Depends(get_current_user),
):
    """Stop a container."""
    service = ContainerService()

    try:
        await service.stop_container(name, timeout=timeout)
        return SuccessResponse(message=f"Container {name} stopped")
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/{name}/restart", response_model=SuccessResponse)
async def restart_container(
    name: str,
    timeout: int = 30,
    _=Depends(get_current_user),
):
    """Restart a container."""
    service = ContainerService()

    try:
        await service.restart_container(name, timeout=timeout)
        return SuccessResponse(message=f"Container {name} restarted")
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.get("/{name}/logs")
async def get_container_logs(
    name: str,
    tail: int = 100,
    since: int = None,
    timestamps: bool = True,
    _=Depends(get_current_user),
):
    """Get container logs."""
    service = ContainerService()

    try:
        logs = await service.get_logs(
            name=name,
            tail=tail,
            since=since,
            timestamps=timestamps,
        )
        return {"container": name, "logs": logs}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.delete("/{name}", response_model=SuccessResponse)
async def remove_container(
    name: str,
    force: bool = False,
    _=Depends(get_current_user),
):
    """Remove a stopped container."""
    service = ContainerService()

    try:
        await service.remove_container(name, force=force)
        return SuccessResponse(message=f"Container {name} removed")
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.post("/{name}/recreate")
async def recreate_container(
    name: str,
    pull: bool = False,
    _=Depends(get_current_user),
):
    """
    Recreate a container using docker compose.

    This will remove the container and create a new one with the same configuration.
    Any non-persisted data will be lost.

    Args:
        name: Container name
        pull: If True, pull the latest image before recreating
    """
    service = ContainerService()

    try:
        result = await service.recreate_container(name, pull=pull)
        return result
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )

File: management/frontend/src/stores/containers.js
/*
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
/management/frontend/src/stores/containers.js

Part of the "n8n_nginx/n8n_management" suite
Version 3.0.0 - January 1st, 2026

Richard J. Sears
richard@n8nmanagement.net
https://github.com/rjsears
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
*/

import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import api from '../services/api'

export const useContainerStore = defineStore('containers', () => {
  // State
  const containers = ref([])
  const stats = ref([])
  const loading = ref(false)
  const refreshing = ref(false) // For force refresh indicator
  const error = ref(null)
  const lastUpdated = ref(null)
  const cacheStatus = ref({ source: null, cached_at: null, age_seconds: null })

  // Getters - with defensive array checks
  const containerList = computed(() =>
    Array.isArray(containers.value) ? containers.value : []
  )

  const runningCount = computed(() =>
    containerList.value.filter(c => c.status === 'running').length
  )

  const stoppedCount = computed(() =>
    containerList.value.filter(c => c.status !== 'running').length
  )

  const unhealthyCount = computed(() =>
    containerList.value.filter(c => c.health === 'unhealthy').length
  )

  const allHealthy = computed(() =>
    containerList.value.length > 0 && containerList.value.every(c =>
      c.status === 'running' && c.health !== 'unhealthy'
    )
  )

  // Actions
  async function fetchContainers(forceRefresh = false) {
    if (forceRefresh) {
      refreshing.value = true
    } else {
      loading.value = true
    }
    error.value = null

    try {
      // Explicitly pass all=true to include stopped containers
      const response = await api.get('/containers/', { params: { all: true, force_refresh: forceRefresh } })
      const data = Array.isArray(response.data) ? response.data : []
      containers.value = data

      // Extract cache status from first item if available
      if (data.length > 0 && data[0]._cached !== undefined) {
        cacheStatus.value = {
          source: 'cache',
          cached_at: data[0]._cached_at,
          age_seconds: data[0]._age_seconds,
        }
      } else {
        cacheStatus.value = { source: 'direct', cached_at: null, age_seconds: null }
      }

      lastUpdated.value = new Date()
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to fetch containers'
      containers.value = []
    } finally {
      loading.value = false
      refreshing.value = false
    }
  }

  async function fetchStats(forceRefresh = false) {
    try {
      const response = await api.get('/containers/stats', { params: { force_refresh: forceRefresh } })
      stats.value = response.data
    } catch (err) {
      console.error('Failed to fetch container stats:', err)
    }
  }

  async function startContainer(name) {
    try {
      await api.post(`/containers/${name}/start`)
      await fetchContainers()
      return true
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to start container'
      return false
    }
  }

  async function stopContainer(name) {
    try {
      await api.post(`/containers/${name}/stop`)
      await fetchContainers()
      return true
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to stop container'
      return false
    }
  }

  async function restartContainer(name) {
    try {
      await api.post(`/containers/${name}/restart`)
      await fetchContainers()
      return true
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to restart container'
      return false
    }
  }

  async function removeContainer(name) {
    try {
      await api.delete(`/containers/${name}`)
      // Small delay to ensure Docker has fully removed the container
      await new Promise(resolve => setTimeout(resolve, 500))
      await fetchContainers()
      return true
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to remove container'
      return false
    }
  }

  async function recreateContainer(name, pull = false) {
    try {
      const response = await api.post(`/containers/${name}/recreate`, null, {
        params: { pull }
      })
      await fetchContainers()
      return response.data
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to recreate container'
      throw err
    }
  }

  async function getLogs(name, options = {}) {
    try {
      const params = {}
      // Support both 'tail' and 'lines' parameter names
      if (options.lines !== undefined) {
        params.tail = options.lines
      } else if (options.tail !== undefined) {
        params.tail = options.tail
      } else {
        params.tail = 100
      }
      // Add since parameter if provided
      if (options.since) {
        params.since = options.since
      }
      const response = await api.get(`/containers/${name}/logs`, { params })
      return response.data.logs
    } catch (err) {
      error.value = err.response?.data?.detail || 'Failed to fetch logs'
      return null
    }
  }

  function getContainerByName(name) {
    return containers.value.find(c => c.name === name)
  }

  function getStatsByName(name) {
    return stats.value.find(s => s.name === name)
  }

  return {
    // State
    containers,
    stats,
    loading,
    refreshing,
    error,
    lastUpdated,
    cacheStatus,
    // Getters
    containerList,
    runningCount,
    stoppedCount,
    unhealthyCount,
    allHealthy,
    // Actions
    fetchContainers,
    fetchStats,
    startContainer,
    stopContainer,
    restartContainer,
    removeContainer,
    recreateContainer,
    getLogs,
    getContainerLogs: getLogs,  // Alias for ContainersView compatibility
    getContainerByName,
    getStatsByName,
  }
})
